docs/src/literate-tutorials/gpu_qp_heat_equation.jl

```diff
@@ -0,0 +1,172 @@
+    n_basefuncs = getnbasefunctions(cellvalues)
+    Ke = zeros(n_basefuncs, n_basefuncs)
+    fe = zeros(n_basefuncs)
+    return (;f, K, assembler, Ke, fe)
+end
+
+
+# Standard global assembly
+
+function assemble_global!(cellvalues, dh::DofHandler,qp_iter::Val{QPiter}) where {QPiter}
+    (;f, K, assembler, Ke, fe) = create_buffers(cellvalues,dh)
+    # Loop over all cels
+    for cell in CellIterator(dh)
+        fill!(Ke, 0)
+        fill!(fe, 0)
+        if QPiter
+            #reinit!(cellvalues, getcoordinates(cell))
+            assemble_element_qpiter!(Ke, fe, cellvalues,getcoordinates(cell))
+        else
+            # Reinitialize cellvalues for this cell
+            reinit!(cellvalues, cell)
+            # Compute element contribution
+            assemble_element_std!(Ke, fe, cellvalues)
+        end
+        # Assemble Ke and fe into K and f
+        @show typeof(cell)
+        assemble!(assembler, celldofs(cell), Ke, fe)
+    end
+    return K, f
+end
+
+
+function assemble_element_gpu_ele_per_thread!(assembler,cv,dh,n_cells_colored,eles_colored)
+    tx = threadIdx().x 
+    bx = blockIdx().x
+    bd = blockDim().x
+    e_color = tx + (bx-1)*bd # element number per color
+    e_color ≤ n_cells_colored || return nothing # e here is the current element index.
+    n_basefuncs = getnbasefunctions(cv)
+    e = eles_colored[e_color]
+    cell_coords = getcoordinates(dh.grid, e)
+
+    ke = MMatrix{4,4,Float32}(undef) # Note: using n_basefuncs instead of 4 will throw an error because this type of dynamisim is not supported in GPU.
+    fill!(ke, 0.0f0)
+    fe = MVector{4,Float32}(undef)
+    fill!(fe, 0.0f0)
+     # Loop over quadrature points
+     for qv in Ferrite.QuadratureValuesIterator(cv,cell_coords)
+        ## Get the quadrature weight
+        dΩ = getdetJdV(qv)
+        ## Loop over test shape functions
+        for i in 1:n_basefuncs
+            δu  = shape_value(qv, i)
+            ∇δu = shape_gradient(qv, i)
+            ## Add contribution to fe
+            fe[i] += δu * dΩ
+            ## Loop over trial shape functions
+            for j in 1:n_basefuncs
+                ∇u = shape_gradient(qv, j)
+                ## Add contribution to Ke
+                ke[i,j] += (∇δu ⋅ ∇u) * dΩ
+            end
+        end
+    end
+
+    ## Assemble Ke into Kgpu ##
+    assemble!(assembler, celldofs(dh,e),SMatrix(ke),SVector(fe)) # when passin mutable objects, throws and error
+
+    return nothing
+end
+
+
+
+
+function assemble_global_gpu_color(cellvalues,dh)
+    #Kgpu =   CUDA.zeros(dh.ndofs.x,dh.ndofs.x)
+    K = create_sparsity_pattern(dh,Float32)
+    Kgpu = GPUSparseMatrixCSC( K.m, K.n, K.colptr |> cu, K.rowval |> cu, K.nzval |> cu)
+    fgpu = CUDA.zeros(ndofs(dh))
+    assembler = start_assemble(Kgpu, fgpu)
+    n_colors = length(colors)
+    for i in 1:n_colors
+        kernel = @cuda launch=false assemble_element_gpu_ele_per_thread!(assembler,cellvalues,dh,length(colors[i]),cu(colors[i]))
```

Comment:
Since I just saw it after Fredrik asked (and Maxi also previously): We can hoist the dof handler adaption out of this loop, so we do not need to move it to GPU every kernel construction.